rcl/include/rcl/wait.h

```diff
@@ -0,0 +1,306 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__WAIT_H_
+#define RCL__WAIT_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "rcl/subscription.h"
+#include "rcl/guard_condition.h"
+#include "rcl/time.h"
+#include "rcl/types.h"
+
+struct rcl_wait_set_impl_t;
+
+/// Container for subscription's, guard condition's, etc to be waited on.
+typedef struct rcl_wait_set_t {
+  /// Storage for subscription pointers.
+  const rcl_subscription_t ** subscriptions;
+  size_t size_of_subscriptions;
+  size_t __current_subscription_offset;
+  /// Storage for guard condition pointers.
+  const rcl_guard_condition_t ** guard_conditions;
+  size_t size_of_guard_conditions;
+  size_t __current_guard_condition_offset;
+  /// Allocator for storage.
+  rcl_allocator_t allocator;
+  /// If set to true, actions like add_subscription will fail until cleared.
+  bool pruned;
+  /// Implementation specific storage.
```

Comment:
It is only forward declared, so it cannot be used from above (the calling code does not know what its size is or what members it has. It is used in the implementation of `rcl`, and so does not have to have anything in it from "below" in `rmw`, but it could. This is simply the pointer to implementation pattern (similar to: https://en.wikipedia.org/wiki/Opaque_pointer) which I'm using here to help with ABI compatibility. Basically it serves as the C equivalent to the private section of a class.